### ISSUE TITLE
test_cleanUpThreadPoolEvenBeforeReactorIsRun relies on reference counting

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4497]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4497|
|Type|defect|
|Created|2010-06-08 13:20:43Z|
|Branch|https://github.com/twisted/twisted/tree/threadpool-gc-4497|

`twisted.internet.test.test_threads.ThreadTestsBuilder.test_cleanUpThreadPoolEvenBeforeReactorIsRun` asserts that the reactor thread pool has been garbage collected immediately after running the code which should drop the reactor's reference to it.

On PyPy (and any runtime without reference counting) the thread pool will only have been collected after the garbage collector runs.

<details><summary>Searchable metadata</summary>

```
trac-id__4497 4497
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_threadpool_gc_4497 branches/threadpool-gc-4497
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests__pypy tests, pypy
time__1276003243000000 1276003243000000
changetime__1276438434000000 1276438434000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29256]) Branching to 'threadpool-gc-4497'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [29257]) gc.collect() before checking that the weakref is dead

refs [#4497](https://github.com/twisted/twisted/issues/4497)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli set owner to @exarkun|
|-|-|

Makes sense, simple change, tests pass.  Kicked off a buildbot build too for lack of things to do in this review.  :)
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/threadpool-gc-4497

Looks fine to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [29298]) Merge threadpool-gc-4497

Author: exarkun
Reviewer: cyli
Fixes: [#4497](https://github.com/twisted/twisted/issues/4497)

Explicitly invoke the garbage collector in the reactor threadpool
cleanup test so that the threadpool has a chance to be collected
even on PyPy (or other runtimes without reference counting) before
the check is made to see if it was collected.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
